fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,121 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  // componentDidMount() {
+  /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+  // }
+
+  componentDidMount() {
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(data => this.setState({ todos: data }));
+  }
+
+  addTodo(newTodo) {
+    const requestConfig = {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo) // This is "serializing" the newTodo, just means turning it into a JSON string
+    };
+
+    fetch('/api/todos', requestConfig)
+      .then(response => response.json())
+      .then(todo => {
+        const full = this.state.todos.concat(todo);
+        this.setState({ todos: full });
+      });
+
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+  }
+
+  toggleCompleted(todoId) {
+
+    const full = this.state.todos;
+    function test(index) {
+      return index.todoId === todoId;
+    }
+    const index = full.findIndex(test);
+    const status = this.state.todos[index].isCompleted;
+    const opposite = !status;
+    const requestConfig = {
+      method: 'PATCH',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(full[index])
+    };
+
+    fetch(`/api/todos/${todoId}`, requestConfig)
+      .then(response => response.json())
+      .then(todo => {
+        const full = this.state.todos;
+        full[index].isCompleted = opposite;
+        this.setState({ todos: full });
+      });
```

Comment:
This doesn't seem like it would update the database correctly, the information being sent to the server won't have the `isCompleted` value toggled. And the state is being mutated in the second `.then`, which is bad. A better approach would be something along these lines:
```JAVASCRIPT
const status = this.state.todos[index].isCompleted;
const body = { isCompleted: !status }
const requestConfig = {
  method: 'PATCH',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(body)
};

fetch(`/api/todos/${todoId}`, requestConfig)
  .then(response => response.json())
  .then(todo => {
    const full = [ ...this.state.todos ]; // Make a copy of the array
    full[index] = todo; // Replace updated todo with todo from the server
    
    this.setState({ todos: full });
  });
```